### PR TITLE
graph/formats/rdf: add equi-canonicalisation function

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

This implementation does not have any of the optimisations mentioned in the paper beyond those required for ergonomics in Go. I intend to add these in a later PR. This intention is the reason for having the methods on the `dfs` pointer type.

/cc and thanks @aidhog

### CODE REVIEW

File: graph/formats/rdf/equi_canonical.go [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+
+// inIF returns whether t is in I or F.
+func inIF(t string, f map[string]bool) bool {
+	return isIRI(t) || f[t]
+}
+
+// dfs is a depth-first search strategy.
+type dfs struct{}
+
+// lean returns a core of the RDF graph f using the given strategy.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply graph/formats/rdf/equi_canonical.go]: Yes, fixed.

**Comment by vladimir-ch** [comment]
@kortschak, a status update: I'm reading `rdf-canonicalisation.pdf`, it'll take some time but eventually I'll get to reviewing this PR.

**Comment by kortschak** [comment]
Thanks

File: graph/formats/rdf/equi_canonical.go [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+
+		edges[triple{sub, prd, obj}] = true
+		edges[triple{sub, prd, "*"}] = true
+		edges[triple{"*", prd, obj}] = true
+		switch {
+		case isBlank(sub) && isBlank(obj):
+			f[sub] = false
+			f[obj] = false
+		case isBlank(sub):
+			if _, ok := f[sub]; !ok {
```
Comment (vladimir-ch) [inline graph/formats/rdf/equi_canonical.go:167]: The translation here from the mathematical set notation to Go is fiendish.

File: graph/formats/rdf/equi_canonical.go [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+	default:
+		return attrI.selfs < attrJ.selfs
+	}
+}
+func (m orderedByCodom) Swap(i, j int) {
+	m.maps[i], m.maps[j] = m.maps[j], m.maps[i]
+	m.attrs[i], m.attrs[j] = m.attrs[j], m.attrs[i]
+}
+
+// join evaluate the given pattern, q, joining with solutions in mu.
```
Comment (vladimir-ch) [inline graph/formats/rdf/equi_canonical.go:521]:
```suggestion
// join evaluates the given pattern, q, joining with solutions in m.
```
?

File: graph/formats/rdf/equi_canonical.go [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+// This is takes only a single mapping and so only works for the DFS strategy.
```
Comment (vladimir-ch) [inline graph/formats/rdf/equi_canonical.go:522]:
```suggestion
// This takes only a single mapping and so only works for the DFS strategy.
```
?

**Review by vladimir-ch — APPROVED** [review]
This was hard to review but must have been even harder to write. Impressive work! I went diligently through the code and compared to the pseudo-code in the paper and as far as I can tell it all looks good to me. I only found some tiny nit in one of the private function comments, you can ignore it if you want to merge. Sorry it took me so long.

File: graph/formats/rdf/equi_canonical.go (hunk 2 above)
Comment (kortschak) [reply graph/formats/rdf/equi_canonical.go:167]: Yep.

File: graph/formats/rdf/equi_canonical.go (hunk 3 above)
Comment (kortschak) [reply graph/formats/rdf/equi_canonical.go:521]: Done

File: graph/formats/rdf/equi_canonical.go (hunk 4 above)
Comment (kortschak) [reply graph/formats/rdf/equi_canonical.go:522]: Done

**Comment by kortschak** [comment]
PTAL

**Comment by kortschak** [comment]
Thank you for the review, I understand how difficult it is to fit these kinds of changes in.

**Review by vladimir-ch — APPROVED** [review]

**Pull request merged**